Bottom: 3ff4aafab7bc51058ad5262731cfa7b9c7138957
Top:    2fdfce83f9ec9cc3a3d1d716e73ce1e0066bc607
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-09-02 11:22:21 +0800

vhost: prefetch vring descriptors

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 4f24b10..e7581f9 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -459,7 +459,8 @@ static void handle_tx(struct vhost_net *net)
 	size_t hdr_size;
 	struct socket *sock;
 	struct vhost_net_ubuf_ref *uninitialized_var(ubufs);
-	bool zcopy, zcopy_used;
+	bool zcopy, zcopy_used, cont;
+	struct vring_desc descs[VHOST_RX_BATCH];
 	int i, batched = VHOST_RX_BATCH;
 
 	mutex_lock(&vq->mutex);
@@ -508,6 +509,17 @@ static void handle_tx(struct vhost_net *net)
 			break;
 		}
 
+		if (cont) {
+			__virito16 ring_head = vhost16_to_cpu(vq, heads[0].id);
+			ret = vhost_copy_from_user(vq, descs,
+						   vq->desc + ring_head,
+						   sizeof descs[0]);
+			if (unlikely(ret)) {
+				vq_err(vq, "Failed to get descriptor\n");
+				break;
+			}
+		}
+
 		for (i = 0; i < avails; i++) {
 			head = __vhost_get_vq_desc(vq, vq->iov,
 						   ARRAY_SIZE(vq->iov),
diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 4a5d1a9..15675de 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2481,6 +2481,7 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	u16 last_avail_idx, last_used_idx, total, copied;
 	__virtio16 avail_idx;
 	struct vring_used_elem __user *used;
+	bool contigious = true;
 	int i;
 
 	if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
@@ -2501,6 +2502,8 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 			return -EFAULT;
 		}
 		last_avail_idx = (last_avail_idx + 1) & (vq->num - 1);
+		if (i > 0 && contigious && heads[i].id != heads[i - 1].id + 1)
+			contigious = false;
 	}
 
 	last_used_idx = vq->last_used_idx & (vq->num - 1);
@@ -2520,6 +2523,8 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 		total -= copied;
 	}
 
+	printk("contigious %d\n", contigious);
+
 	/* Only get avail ring entries after they have been exposed by guest. */
 	smp_rmb();
 	return ret;
